### PR TITLE
Exception handling fix in `math_ops.div_no_nan()`

### PR DESCRIPTION
**Pull request by kmh4321**
Currently `div_no_nan()` does not raise a `TypeError` if the inputs are of different types. Although the code for that exists, it is never executed as in one of the lines above that, the code fails with a different error. This is fixed in the first [commit](https://github.com/tensorflow/tensorflow/commit/25ffbdfccc02b7f8c2e05f76c066ddef2503af39). 

Code to execute:

```python
import tensorflow as tf
a = tf.constant(1, dtype=tf.float32)
b = tf.constant(2, dtype=tf.float64)
c = tf.div_no_nan(a,b)
```
Earlier:
```python

ValueError: Tensor conversion requested dtype float32 for Tensor with dtype float64: 'tf.Tensor(2.0, shape=(), dtype=float64)'

```
Now:
```python

TypeError: x and y must have the same dtype, got tf.float32 != tf.float64
```

### CODE REVIEW

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -1105,11 +1105,14 @@ def div_no_nan(x, y, name=None):
...
     The element-wise value of the x divided by y.
+  
+  Raises:
+    TypeError: if x and y are not of the same dtype.
   """
 
   with ops.name_scope(name, "div_no_nan", [x, y]) as name:
     x = ops.convert_to_tensor(x, name="x")
-    y = ops.convert_to_tensor(y, name="y", dtype=x.dtype.base_dtype)
+    y = ops.convert_to_tensor(y, name="y")
```
Comment (alextp) [inline tensorflow/python/ops/math_ops.py]: I don't like this change as it makes code which now does not raise exceptions raise exceptions, like tf.div_no_nan(tf.constant(1.0), 2)

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/math_ops.py (hunk 1 above)
Comment (kmh4321) [reply tensorflow/python/ops/math_ops.py]: While the example you mention above passes without raising exception in the current code, the current code fails if you just swap the order or change the combination and the exceptions raised are inconsistent:

Change in order:
```python 
tf.div_no_nan(2, tf.constant(1.0))
ValueError: Tensor conversion requested dtype int32 for Tensor with dtype float32: <tf.Tensor 'Const_2:0' shape=() dtype=float32>
```

 Change in combination:
```python

tf.div_no_nan(tf.constant(1), 2.0)
TypeError: Expected int32, got 2.0 of type 'float' instead.

```

The minor change in the PR will unify all these exceptions and make it consistent.

Comment (kmh4321) [reply tensorflow/python/ops/math_ops.py]: In the meanwhile I can work on the cases where we pass one tensor (of float) and one int or vice versa can be handled with some other logic without raising exceptions.

Comment (alextp) [reply tensorflow/python/ops/math_ops.py]: Sadly raising exceptions for code which currently does not violates TF's backward compatibility policy so we can't accept this part of the change.

Comment (kmh4321) [reply tensorflow/python/ops/math_ops.py]: but this IS an issue right? There are 3 different exceptions being raised for basically the same input:
`TypeError` and `ValueError` being raised with varying error messages, sometimes from inner segments of code which the end user need not necessarily be exposed to. For an end user in a production system, catching these exceptions will be cumbersome and can break things. 

As for the backward compatibility, the `TypeError` currently being "raised" will never get raised in any scenario so might as well remove it then?

Comment (alextp) [reply tensorflow/python/ops/math_ops.py]: I agree we need to unify things. Happy to remove the raises annotation if it doesn't apply.

Comment (kmh4321) [reply tensorflow/python/ops/math_ops.py]: Removed the exception.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by alextp — APPROVED** [review]

**Pull request merged**